BluetoothExplorer/Controller/GATT Characteristic Editors/DateTimeCharacteristicViewController.swift

```diff
@@ -0,0 +1,126 @@
+import Bluetooth
+import GATT
+
+final class DateTimeCharacteristicViewController: UIViewController {
+    
+    // MARK: - IB Outlets
+    
+    @IBOutlet private(set) weak var dateTimeLabel: UILabel!
+    
+    @IBOutlet private(set) weak var datePicker: UIDatePicker!
+    
+    // MARK: - Properties
+    
+    var value = GATTDateTime(date: Date())
+    
+    var valueDidChange: ((GATTDateTime) -> ())?
+    
+    var minimumDate: Date? {
+        
+        var dateComponents = DateComponents()
+        dateComponents.year = Int(GATTDateTime.Year.min.rawValue)
+        dateComponents.timeZone = TimeZone(identifier: "UTC")
+        
+        return Calendar.current.date(from: dateComponents)
+    }
+    
+    var maximumDate: Date? {
+        
+        var dateComponents = DateComponents()
+        dateComponents.year = Int(GATTDateTime.Year.max.rawValue)
+        dateComponents.timeZone = TimeZone(identifier: "UTC")
+        
+        return Calendar.current.date(from: dateComponents)
+    }
+    
+    // MARK: - Loading
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        guard let minimumDate = minimumDate
+            else { assertionFailure("Couldn't create minimumDate"); return }
+        
+        guard let maximumDate = maximumDate
+            else { assertionFailure("Couldn't create maximumDate"); return }
+        
+        datePicker.minimumDate = minimumDate
+        datePicker.maximumDate = maximumDate
+        
+        configureView()
+    }
+    
+    // MARK: - Actions
+    
+    @IBAction func datePickerChanged(_ sender: Any) {
+        
+        let strDate = CustomDateFormatter.default.string(from: datePicker.date)
+        dateTimeLabel.text = strDate
+        
+        let dateComponents = Calendar.current.dateComponents(in: TimeZone(identifier: "UTC")!, from: datePicker.date)
+        
+        let datetime = GATTDateTime(dateComponents: dateComponents)
+        
+        guard let value = datetime
+            else { assertionFailure("Couldn't create GATTDateTime from components"); return }
+        
+        self.value = value
+        valueDidChange?(value)
+    }
+    
+    // MARK: - Methods
+    
+    func configureView() {
+        
+        guard isViewLoaded else { return }
+        
+        updateDatePicker()
+    }
+    
+    func updateDatePicker() {
+        
+        if let date = value.dateComponents.date {
+            dateTimeLabel.text = CustomDateFormatter.default.string(from: date)
+            datePicker.date = date
+        }
+    }
+    
+}
+
+// MARK: - CharacteristicViewController
+
+extension DateTimeCharacteristicViewController: CharacteristicViewController {
+    
+    static func fromStoryboard() -> DateTimeCharacteristicViewController {
+        
+        let storyboard = UIStoryboard(name: "DateTimeCharacteristic", bundle: .main)
+        
+        let viewController = storyboard.instantiateInitialViewController() as! DateTimeCharacteristicViewController
```

Comment:
I am open to any suggestions for improvement or code reusability for this.

Reply:
I like to use this approach

protocol InstantiableController {
    
    static func instantiate<T>() -> T
}

extension InstantiableController where Self: UIViewController {
    
    static func instantiate<T>() -> T {
        guard let storyboardName = String(describing: self).text(before: "ViewController") else {
            fatalError("The controller name is not standard.")
        }
        
        let storyboard = UIStoryboard(name: storyboardName, bundle: Bundle(for: self))
        let identifier = String(describing: T.self)
        guard let viewController = storyboard.instantiateViewController(withIdentifier: identifier) as? T else {
            fatalError("The storyboard identifier does not exist.")
        }
        
        return viewController
    }
    
}

For example:
`Storyboard`: DateTimeCharacteristic
`ViewController`: DateTimeCharacteristicViewController

where the name of the controller must be the same as the storyboard plus `ViewController`

To avoid mistakes on storyboard and controller naming, I use unit tests.

In that way I just need to add `InstantiableController` to any controller and call:

let dateTimeController: DateTimeCharacteristicViewController = .instantiate()

Reply:
Use that approach then, make it a separate protocol.